File: painel/src/zonas.ts
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Quero Automação Ltda

// Why: a zone is a multiroom equipment occupying one of the six blocks of section 8, and
// nothing here decides which equipment that may be nor which data point a block carries: the
// daemon answers both, so the panel never keeps a second copy of the contract. What lives
// here is what a screen needs from that answer, as pure functions with tests.
// Por que: uma zona é um equipamento multiroom ocupando um dos seis blocos da seção 8, e nada
// aqui decide qual equipamento pode ser nem qual data point um bloco carrega: o daemon
// responde os dois, então o painel nunca guarda uma segunda cópia do contrato. Aqui mora o
// que uma tela precisa daquela resposta, como funções puras com teste.

import {
  lerEstadoEquipamento,
  lerLista,
  type Capacidade,
  type EstadoEquipamento,
  type ItemCatalogo,
} from "./equipamentos.ts";

export const SOLO = "solo";
export const PREFIXO_GRUPO = "grupo";

// Why: section 6 fixes what a zone is, and the daemon refuses a block for anything else; the
// panel offers the same set so the integrator is never offered a refusal.
// Por que: a seção 6 fixa o que é uma zona, e o daemon recusa um bloco para qualquer outra
// coisa; o painel oferece o mesmo conjunto para o integrador nunca receber uma recusa.
export const CATEGORIA_DE_ZONA = "multiroom";
export const CAPACIDADE_DE_GRUPO = "agrupar";

export const PAPEIS = ["", "mestre", "escravo"] as const;
export type Papel = (typeof PAPEIS)[number];

// The functions of a block, in the order section 8 numbers them.
// As funções de um bloco, na ordem em que a seção 8 as numera.
export const FUNCOES_DO_BLOCO = [
  "volume",
  "play",
  "preset",
  "online",
  "tocando",
  "entrada",
] as const;

export type FuncaoDoBloco = (typeof FUNCOES_DO_BLOCO)[number];

// Why: the API answers a stable code and never a phrase, so each one needs an entry in both
// dictionaries, and a test asserts that it has one.
// Por que: a API responde um código estável e nunca uma frase, então cada um precisa de
// entrada nos dois dicionários, e um teste garante isso.
export const CODIGOS_ZONAS = [
  "zonas_demais",
  "zona_repetida",
  "eq_nao_multiroom",
  "identidade_invalida",
  "dp_desconhecido",
  "dp_somente_leitura",
  "valor_invalido",
  "zona_offline",
] as const;

export type DpsDoBloco = Record<FuncaoDoBloco, number>;

export interface Zona {
  zona: number;
  identidade: string;
  nome: string;
  tipo: string;
  papel: Papel;
  entradas: string[];
  dps: DpsDoBloco;
  estado: EstadoEquipamento | null;
}

export interface LeituraDeZonas {
  zonas: Zona[];
  grupo: string;
  dp_grupo: number;
}

type Objeto = Record<string, unknown>;

const ehNumero = (valor: unknown): valor is number => typeof valor === "number";
const ehTexto = (valor: unknown): valor is string => typeof valor === "string";

function ehObjeto(valor: unknown): valor is Objeto {
  return typeof valor === "object" && valor !== null && !Array.isArray(valor);
}

function lerPapel(valor: unknown): Papel | null {
  return PAPEIS.find((papel) => papel === valor) ?? null;
}

function lerDps(valor: unknown): DpsDoBloco | null {
  if (!ehObjeto(valor)) return null;
  const dps: Partial<DpsDoBloco> = {};
  for (const funcao of FUNCOES_DO_BLOCO) {
    const dpid = valor[funcao];
    if (!ehNumero(dpid)) return null;
    dps[funcao] = dpid;
  }
  return dps as DpsDoBloco;
}

export function lerZona(valor: unknown): Zona | null {
  if (!ehObjeto(valor) || !ehNumero(valor.zona)) return null;
  const { identidade, nome, tipo } = valor;
  if (!ehTexto(identidade) || !ehTexto(nome) || !ehTexto(tipo)) return null;
  const papel = lerPapel(valor.papel);
  const entradas = lerLista(valor.entradas, (bruto) => (ehTexto(bruto) ? bruto : null));
  const dps = lerDps(valor.dps);
  if (papel === null || entradas === null || dps === null) return null;
  // Why: a block nobody occupies answers a null state, which is not a broken answer: the hub
  // works with zero equipment and the POSITION of the block is the contract.
  // Por que: um bloco que ninguém ocupa responde estado nulo, o que não é resposta quebrada: o
  // hub funciona com zero equipamento e a POSIÇÃO do bloco é o contrato.
  const estado = valor.estado === null ? null : lerEstadoEquipamento(valor.estado);
  if (estado === undefined) return null;
  return { zona: valor.zona, identidade, nome, tipo, papel, entradas, dps, estado };
}

export function lerLeituraDeZonas(dados: Objeto): LeituraDeZonas | null {
  const zonas = lerLista(dados.zonas, lerZona);
  if (zonas === null || !ehTexto(dados.grupo) || !ehNumero(dados.dp_grupo)) return null;
  return { zonas, grupo: dados.grupo, dp_grupo: dados.dp_grupo };
}

export function valorDoGrupo(zona: number): string {
  return `${PREFIXO_GRUPO}${zona}`;
}

export function ordemDe(zonas: readonly Zona[]): string[] {
  return zonas.map((bloco) => bloco.identidade);
}

// Why: a shift would move the speaker of zone 2 into zone 1 in every automation the customer
// already built, so a block is emptied where it is and the identity is only taken off the
// block it used to occupy.
// Por que: um empurrão moveria a caixa da zona 2 para a zona 1 em toda automação que o cliente
// já montou, então um bloco é esvaziado no lugar e a identidade só sai do bloco que ela
// ocupava.
export function comIdentidade(
  ordem: readonly string[],
  zona: number,
  identidade: string,
): string[] {
  return ordem.map((atual, posicao) => {
    if (posicao === zona - 1) return identidade;
    return identidade !== "" && atual === identidade ? "" : atual;
  });
}

export function podeOcuparBloco(item: ItemCatalogo | undefined): boolean {
  if (item === undefined) return false;
  return (
    item.categoria === CATEGORIA_DE_ZONA &&
    item.capacidades.includes(CAPACIDADE_DE_GRUPO as Capacidade)
  );
}

// Why: section 14, a group only ever exists between speakers of the same domain, so a zone
// that has nobody of its own tipo to lead is never offered as a group; offering a mixed one
// is what leaves half of it playing and the other half silent.
// Por que: seção 14, um grupo só existe entre caixas do mesmo domínio, então uma zona que não
// tem ninguém do tipo dela para liderar nunca é oferecida como grupo; oferecer um misto é o
// que deixa metade dele tocando e a outra metade calada.
export function gruposPossiveis(zonas: readonly Zona[], catalogo: readonly ItemCatalogo[]): string[] {
  const item = (tipo: string): ItemCatalogo | undefined =>
    catalogo.find((candidato) => candidato.tipo === tipo);
  const lideres = zonas.filter(
    (bloco) =>
      bloco.identidade !== "" &&
      podeOcuparBloco(item(bloco.tipo)) &&
      zonas.some(
        (outra) =>
          outra.zona !== bloco.zona && outra.identidade !== "" && outra.tipo === bloco.tipo,
      ),
  );
  return [SOLO, ...lideres.map((bloco) => valorDoGrupo(bloco.zona))];
}

export type EspecieDeControle = "escala" | "alternar" | "escolha" | "preset";

export interface ControleDeZona {
  funcao: FuncaoDoBloco;
  dpid: number;
  especie: EspecieDeControle;
}

// Why: section 6, a capability the manifest does not declare gets no button, because the
// daemon answers nao_suportado before the driver is touched. The play data point is a toggle
// over the two transport capabilities, so a driver that declares only one of them gets
// neither half of the button instead of a button that fails every other press.
// Por que: seção 6, uma capacidade que o manifesto não declara não ganha botão, porque o
// daemon responde nao_suportado antes de tocar no driver. O data point de play é uma chave
// sobre as duas capacidades de transporte, então um driver que declara só uma delas não ganha
// metade nenhuma do botão em vez de um botão que falha a cada duas apertadas.
const EXIGIDAS: Record<FuncaoDoBloco, Capacidade[]> = {
  volume: ["volume"],
  play: ["tocar", "pausar"],
  preset: ["comando_extra"],
  entrada: ["fonte"],
  online: [],
  tocando: [],
};

const ESPECIES: Partial<Record<FuncaoDoBloco, EspecieDeControle>> = {
  volume: "escala",
  play: "alternar",
  preset: "preset",
  entrada: "escolha",
};

export function controlesDaZona(
  zona: Zona,
  item: ItemCatalogo | undefined,
): ControleDeZona[] {
  if (zona.identidade === "" || item === undefined) return [];
  const controles: ControleDeZona[] = [];
  for (const funcao of FUNCOES_DO_BLOCO) {
    const especie = ESPECIES[funcao];
    const exigidas = EXIGIDAS[funcao];
    if (especie === undefined || exigidas.length === 0) continue;
    if (!exigidas.every((capacidade) => item.capacidades.includes(capacidade))) continue;
    if (funcao === "entrada" && zona.entradas.length === 0) continue;
    controles.push({ funcao, dpid: zona.dps[funcao], especie });
  }
  return controles;
}

export type Preparo = { ok: true; valor: unknown } | { ok: false; codigo: string };

// Why: a volume of 300 is refused here with the same stable code the daemon would answer, so
// a typo costs no request; the daemon still judges it, because the panel is not the authority.
// Por que: um volume de 300 é recusado aqui com o mesmo código estável que o daemon
// responderia, então um erro de digitação não custa requisição; o daemon ainda julga, porque o
// painel não é a autoridade.
export function prepararVolume(entrada: string): Preparo {
  const limpo = entrada.trim();
  const dentro = /^\d{1,3}$/.test(limpo) && Number(limpo) <= 100;
  return dentro ? { ok: true, valor: Number(limpo) } : { ok: false, codigo: "valor_invalido" };
}

export function tocando(zona: Zona): boolean {
  // Why: section 14, a slave answers stop even while the group plays, so the daemon mirrors
  // what the master plays onto it and the screen reads that and never the slave itself.
  // Por que: seção 14, um escravo responde stop mesmo com o grupo tocando, então o daemon
  // espelha nele o que o mestre toca e a tela lê isso, e nunca o próprio escravo.
  return zona.estado !== null && zona.estado.tocando !== null && zona.estado.tocando !== "";
}

File: core/iphub/api/zonas.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (C) 2026 Quero Automação Ltda
"""Section 8 over HTTP: the order of the six blocks, the data points and the group.

A zone is what section 6 says it is, a multiroom equipment occupying one of the six blocks,
so these routes carry no second registry: the order is a list of identities already
registered as equipment, and the module of the zones is the one that judges it and answers a
stable code. Nothing here decides where a set goes either; aplicar_dp of the common module
does, so the panel and the bus of the same hub cannot disagree about it.

The numbers of section 8 leave the daemon in the answers instead of being written a second
time in the panel: each block carries the data point of each of its functions, and the
snapshot route carries the whole table. A panel that computed the numbering would be a
second copy of the contract, and the day the contract moved only one of them would move.

Seção 8 sobre HTTP: a ordem dos seis blocos, os data points e o grupo.

Uma zona é o que a seção 6 diz que ela é, um equipamento multiroom ocupando um dos seis
blocos, então estas rotas não carregam segundo cadastro: a ordem é uma lista de identidades
já cadastradas como equipamento, e o módulo das zonas é quem a julga e responde um código
estável. Nada aqui decide para onde vai um set tampouco; quem decide é o aplicar_dp do módulo
comum, para o painel e o barramento do mesmo hub não poderem discordar disso.

Os números da seção 8 saem do daemon nas respostas em vez de serem escritos uma segunda vez
no painel: cada bloco carrega o data point de cada função dele, e a rota de snapshot carrega
a tabela inteira. Um painel que calculasse a numeração seria uma segunda cópia do contrato, e
no dia em que o contrato mudasse só uma das duas mudaria.
"""

import logging
import re
from dataclasses import replace

from aiohttp import web

from iphub import cenas
from iphub.api.comum import (
    GESTOR,
    aplicar_dp,
    com_sessao,
    config_de,
    ler_corpo,
    resposta_ok,
    trocar_config,
    valores_dps,
    zonas_de,
)
from iphub.api.formato import estado_json
from iphub.config import Cadastro
from iphub.dpbus import mapa, protocolo
from iphub.dpbus import zonas as modulo
from iphub.drivers.manifesto import Estado
from iphub.portao import resposta_erro

log = logging.getLogger("iphub.api.zonas")

_SO_DIGITOS = re.compile(r"[0-9]{1,10}")

CORPO_INVALIDO = "corpo_invalido"
ERRO_INTERNO = "erro_interno"

PAPEL_MESTRE = "mestre"
PAPEL_ESCRAVO = "escravo"
PAPEL_SOLO = ""

# The status of every stable code these routes answer; nothing else reaches the panel.
# O status de todo código estável que estas rotas respondem; nada mais chega ao painel.
STATUS_POR_CODIGO = {
    CORPO_INVALIDO: 400,
    modulo.ZONAS_DEMAIS: 400,
    modulo.ZONA_REPETIDA: 400,
    modulo.EQ_NAO_MULTIROOM: 400,
    modulo.IDENTIDADE_INVALIDA: 400,
    "eq_nao_encontrado": 404,
    protocolo.DP_DESCONHECIDO: 404,
    protocolo.DP_SOMENTE_LEITURA: 400,
    protocolo.VALOR_INVALIDO: 400,
    protocolo.ZONA_OFFLINE: 503,
    "nao_suportado": 400,
    "auth_pendente": 409,
    "erro_aparelho": 502,
    # Why: DP 131 runs a scene, so the two codes the scene executor answers reach this route
    # too, and without them the panel got a 500 with erro_interno for a scene that simply does
    # not exist or is already running, while the scenes route answered 404 and 409 for the
    # very same conditions.
    # Por que: o DP 131 executa uma cena, então os dois códigos que o executor de cenas
    # responde chegam também a esta rota, e sem eles o painel recebia um 500 com erro_interno
    # para uma cena que simplesmente não existe ou já está rodando, enquanto a rota de cenas
    # respondia 404 e 409 para as mesmíssimas condições.
    cenas.CENA_NAO_ENCONTRADA: 404,
    cenas.CENA_EM_CURSO: 409,
    ERRO_INTERNO: 500,
}


def erro(codigo: str) -> web.Response:
    """The stable code as the status the panel reads, and never a status invented here.

    O código estável como o status que o painel lê, e nunca um status inventado aqui.
    """
    return resposta_erro(STATUS_POR_CODIGO.get(codigo, 500), codigo)


def _cadastros(request: web.Request) -> dict[str, Cadastro]:
    return {cadastro.identidade: cadastro for cadastro in request.app[GESTOR].cadastros}


def _papel(zona: int, mestre: int, escravos: tuple[int, ...], alheios: tuple[int, ...]) -> str:
    """What the speaker of this block really is right now, and not only what the hub asked.

    Why: a speaker the customer grouped with the app of the manufacturer is a slave of a group
    this hub does not lead, and it refuses volume, transport, preset and input. Calling that
    solo drew a panel full of controls that only ever answer no, with nothing saying why.

    O que a caixa deste bloco realmente é agora, e não só o que o hub pediu.

    Por que: uma caixa que o cliente agrupou com o app do fabricante é escrava de um grupo que
    este hub não lidera, e ela recusa volume, transporte, preset e entrada. Chamar isso de solo
    desenhava um painel cheio de controles que só respondem não, sem nada dizendo por quê.
    """
    if mestre and zona == mestre:
        return PAPEL_MESTRE
    if zona in escravos or zona in alheios:
        return PAPEL_ESCRAVO
    return PAPEL_SOLO


def _entradas(zonas: modulo.Zonas, zona: int) -> tuple[str, ...]:
    """The inputs the bus really takes for one block, which the hardware decides.

    As entradas que o barramento realmente aceita para um bloco, que o hardware decide.
    """
    for dp in mapa.da_zona(zona):
        if dp.funcao == mapa.FUNCAO_ENTRADA:
            return zonas.valores_de(dp)
    return ()


def _dps_do_bloco(zona: int) -> dict[str, int]:
    """The data point of every function of one block, straight from the map of section 8.

    O data point de cada função de um bloco, direto do mapa da seção 8.
    """
    return {funcao: mapa.dp_de(zona, funcao) for funcao in mapa.FUNCOES_ZONA}


def _zona_json(
    zona: int,
    cadastro: Cadastro | None,
    estado: Estado | None,
    entradas: tuple[str, ...],
    papel: str,
) -> dict:
    """One block as the panel reads it: who occupies it, what it says and what it takes.

    Um bloco como o painel o lê: quem o ocupa, o que ele diz e o que ele aceita.
    """
    # Why: a block nobody occupies is a normal state of the hub (section 6 works with zero
    # equipment), so it answers with an empty identity and a null state instead of vanishing
    # from the list; the POSITION is the contract, and a shorter list would move it.
    # Por que: um bloco que ninguém ocupa é estado normal do hub (a seção 6 funciona com zero
    # equipamento), então ele responde com identidade vazia e estado nulo em vez de sumir da
    # lista; a POSIÇÃO é o contrato, e uma lista mais curta a moveria.
    return {
        "zona": zona,
        "identidade": "" if cadastro is None else cadastro.identidade,
        "nome": "" if cadastro is None else cadastro.nome,
        "tipo": "" if cadastro is None else cadastro.tipo,
        "papel": papel,
        "entradas": list(entradas),
        "dps": _dps_do_bloco(zona),
        "estado": None if estado is None else estado_json(estado),
    }


@com_sessao
async def listar(request: web.Request) -> web.Response:
    zonas = zonas_de(request.app)
    cadastros = _cadastros(request)
    estados = request.app[GESTOR].estados()
    mestre = modulo.zona_do_grupo(zonas.grupo()) or 0
    escravos = zonas.escravos()
    alheios = zonas.escravos_alheios()
    blocos = []
    for zona in range(1, mapa.ZONAS + 1):
        identidade = zonas.identidade(zona)
        blocos.append(
            _zona_json(
                zona,
                cadastros.get(identidade),
                estados.get(identidade),
                _entradas(zonas, zona),
                _papel(zona, mestre, escravos, alheios),
            )
        )
    return resposta_ok(zonas=blocos, grupo=zonas.grupo(), dp_grupo=mapa.GRUPO)


@com_sessao
async def definir(request: web.Request) -> web.Response:
    """Saves the order of the blocks, which is the only registry a zone ever has.

    Grava a ordem dos blocos, que é o único cadastro que uma zona tem.
    """
    app = request.app
    dados = await ler_corpo(request)
    if dados is None:
        return erro(CORPO_INVALIDO)
    zonas = zonas_de(app)
    try:
        ordem = zonas.validar(dados.get("zonas"))
    except modulo.OrdemInvalida as recusa:
        return erro(recusa.codigo)
    # Why: the file is written before the order is applied, because an order that lived only
    # in memory would move a speaker back to its old block on the next boot, in silence, on a
    # bus the customer already automated.
    # Por que: o arquivo é gravado antes de a ordem valer, porque uma ordem que vivesse só na
    # memória moveria uma caixa de volta ao bloco antigo no próximo boot, em silêncio, num
    # barramento que o cliente já automatizou.
    try:
        trocar_config(app, replace(config_de(app), zonas=ordem))
    except OSError as falha:
        log.error("could not write the order of the zones: %s", falha)
        return erro(ERRO_INTERNO)
    try:
        # Why: the module judges the order again under its own lock, so an equipment removed
        # between the two calls is a stable code and never a 500 with a traceback.
        # Por que: o módulo julga a ordem de novo sob a trava dele, então um equipamento
        # removido entre as duas chamadas é um código estável e nunca um 500 com traceback.
        await zonas.definir_ordem(ordem)
    except modulo.OrdemInvalida as recusa:
        return erro(recusa.codigo)
    return resposta_ok(zonas=list(ordem))


@com_sessao
async def dps(request: web.Request) -> web.Response:
    """Every reportable data point and the table of section 8, for the panel and for curl.

    Todo data point reportável e a tabela da seção 8, para o painel e para o curl.
    """
    quadro = protocolo.snapshot(valores_dps(request.app))
    zonas = zonas_de(request.app)
    tabela = [
        {
            "dpid": dp.dpid,
            "zona": dp.zona,
            "funcao": dp.funcao,
            "tipo": dp.tipo.value,
            "sentido": dp.sentido.value,
            # Why: the inputs of a speaker come from the hardware (section 14, plm_support),
            # so the values of that enum are the ones this hub really offers right now and
            # not a list the panel could guess from the contract.
            # Por que: as entradas de uma caixa vêm do hardware (seção 14, plm_support), então
            # os valores daquele enum são os que este hub realmente oferece agora e não uma
            # lista que o painel pudesse adivinhar do contrato.
            "valores": list(zonas.valores_de(dp)),
        }
        for dp in mapa.DPS
    ]
    return resposta_ok(dps=quadro["dps"], mapa=tabela)


@com_sessao
async def ajustar(request: web.Request) -> web.Response:
    """One set of section 8 by HTTP, with the same effect a set on the bus would have.

    Um set da seção 8 por HTTP, com o mesmo efeito que um set no barramento teria.
    """
    dados = await ler_corpo(request)
    if dados is None:
        return erro(CORPO_INVALIDO)
    codigo = await aplicar_dp(request.app, _dpid(request), dados.get("v"))
    return resposta_ok() if codigo is None else erro(codigo)


@com_sessao
async def grupo(request: web.Request) -> web.Response:
    """DP 132 by name: solo takes the group down, grupoN forms the one that zone leads.

    O DP 132 pelo nome: solo derruba o grupo, grupoN forma o que aquela zona lidera.
    """
    dados = await ler_corpo(request)
    if dados is None:
        return erro(CORPO_INVALIDO)
    codigo = await aplicar_dp(request.app, mapa.GRUPO, dados.get("v"))
    return resposta_ok(grupo=zonas_de(request.app).grupo()) if codigo is None else erro(codigo)


def _dpid(request: web.Request) -> object:
    """The number in the path, or the raw text when it is not a number at all.

    O número no caminho, ou o texto cru quando ele não é número nenhum.
    """
    # Why: the map answers dp_desconhecido for anything that is not an int, so a path of
    # letters travels as text and is refused by the same rule that refuses dp 999.
    # Por que: o mapa responde dp_desconhecido para o que não for int, então um caminho de
    # letras viaja como texto e é recusado pela mesma regra que recusa o dp 999.
    bruto = request.match_info["dpid"]
    # Why: str.isdigit() is true for characters int() refuses, such as the superscript two,
    # so a session holder could turn this route into a 500 with a traceback in the log. Only
    # an ASCII digit is a number here, and the ceiling keeps a path of a thousand digits from
    # becoming a thousand digit int.
    # Por que: str.isdigit() é verdadeiro para caracteres que o int() recusa, como o dois
    # sobrescrito, então quem tem sessão podia transformar esta rota num 500 com traceback no
    # log. Só dígito ASCII é número aqui, e o teto impede que um caminho de mil dígitos vire
    # um int de mil dígitos.
    return int(bruto) if _SO_DIGITOS.fullmatch(bruto) else bruto
